### PR TITLE
Get All feature flags

### PR DESCRIPTION
**Pull request by vikhyat187**
Tasks to be done in future.
- [ ] Adding caching to the API.
- [ ] Find other ways to optimise this API.

REF : https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html

### CODE REVIEW

File: getAllFeatureFlags/main.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		return nil, err
+	}
+
+	return featureFlagsResponse, nil
+}
+
+func handler(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
+	db := database.CreateDynamoDB()
+
+	log.Println(db, " is the database")
```
Comment (shubham-y) [inline getAllFeatureFlags/main.go:46]: Please remove this unwanted log

File: getAllFeatureFlags/main.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+	if err != nil {
+		return utils.ServerError(err)
+	}
+
+	if featureFlagsResponse == nil {
+		return events.APIGatewayProxyResponse{
+			Body: "No entries found.",
+			StatusCode : http.StatusNotFound,
+		}, nil
```
Comment (shubham-y) [inline getAllFeatureFlags/main.go:57]: It will be better to use the `utils.ClientError` function here

File: getAllFeatureFlags/main.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	jsonResult, err := json.Marshal(featureFlagsResponse)
+	if err != nil {
+		log.Println("Error converting feature flags to JSON")
+		utils.ServerError(err)
```
Comment (shubham-y) [inline getAllFeatureFlags/main.go:63]: Please add `return` here

File: getAllFeatureFlags/main.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+	"log"
+	"encoding/json"
+	"net/http"
+)
+
+func getAllFeatureFlags(db *dynamodb.DynamoDB) ([]utils.FeatureFlagResponse, error) {
+
+	input := &dynamodb.ScanInput{
+		TableName: aws.String(database.GetFeatureFlagTableName()),
```
Comment (shubham-y) [inline getAllFeatureFlags/main.go:19]: Please use the `GetTableName()` instead
https://github.com/Real-Dev-Squad/feature-flag-backend/blob/ea0e2a48cdecd926bc615308e56c03ec1c972436/database/dynamodb.go#L61-L71

File: getAllFeatureFlags/main.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+	jsonResult, err := json.Marshal(featureFlagsResponse)
+	if err != nil {
+		log.Println("Error converting feature flags to JSON")
+		utils.ServerError(err)
+	}
+
+	return events.APIGatewayProxyResponse{
+		Body:      	string(jsonResult),
+		StatusCode: http.StatusOK,
```
Comment (shubham-y) [inline getAllFeatureFlags/main.go:68]: NIT: formatting issue
```suggestion
		Body:       string(jsonResult),
		StatusCode: http.StatusOK,
```

File: getAllFeatureFlags/go.mod [hunk 6]
```diff
@@ -0,0 +1,14 @@
+module getAllFeatureFlags
```
Comment (shubham-y) [inline getAllFeatureFlags/go.mod:1]: The entry for this function is missing in `template.yaml` file

Comment (vikhyat187) [reply getAllFeatureFlags/go.mod:1]: done

File: getAllFeatureFlags/main.go (hunk 4 above)
Comment (vikhyat187) [reply getAllFeatureFlags/main.go]: updated

File: getAllFeatureFlags/main.go (hunk 2 above)
Comment (vikhyat187) [reply getAllFeatureFlags/main.go]: done

File: getAllFeatureFlags/main.go [hunk 7]
```diff
@@ -43,28 +43,24 @@ func getAllFeatureFlags(db *dynamodb.DynamoDB) ([]utils.FeatureFlagResponse, err
...
 	if err != nil {
 		return utils.ServerError(err)
 	}
 
 	if featureFlagsResponse == nil {
-		return events.APIGatewayProxyResponse{
-			Body: "No entries found.",
-			StatusCode : http.StatusNotFound,
-		}, nil
+		utils.ClientError(http.StatusNotFound, "No feature flags found :( ")
```
Comment (shubham-y) [inline getAllFeatureFlags/main.go:53]: Please add `return` here

Comment (shubham-y) [reply getAllFeatureFlags/main.go:53]: Also I think the error message `No feature flags found` is better

File: getAllFeatureFlags/main.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	result, err := db.Scan(input)
+
+	if err != nil {
+		utils.DdbError(err)
+		return nil, err
+	}
+
+	if len(result.Items) == 0 {
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (vikhyat187) [reply getAllFeatureFlags/main.go:29]: Sure, inline.

File: getAllFeatureFlags/main.go (hunk 7 above)
Comment (vikhyat187) [reply getAllFeatureFlags/main.go:53]: Sure.

File: getAllFeatureFlags/main.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (vikhyat187) [reply getAllFeatureFlags/main.go:29]: @RitikJaiswal75 I was going through one of the resource REF: https://morioh.com/p/385f6171e967
I was thinking should I go with `nilSlice` or the `emptySlice`, I'm aligned towards `emptySlice` what are your thoughts?

[the remark this answers is not in the document]
Comment (vikhyat187) [reply getAllFeatureFlags/main.go]: Updated this @RitikJaiswal75 please have a look

File: getAllFeatureFlags/main.go (hunk 7 above)
Comment (vikhyat187) [reply getAllFeatureFlags/main.go]: done resolving this

**Review by shubham-y — APPROVED** [review]

**Pull request merged**